rest/multipart.go

```diff
@@ -59,12 +52,85 @@ func ReadJSONFromMIME(headers http.Header, input io.ReadCloser, into interface{}
 			err = base.HTTPErrorf(http.StatusBadRequest, "Bad JSON: %s", err.Error())
 		}
 	}
+	return err
+}
+
+// ReadSanitizeConfigJSON parses a JSON body, unmarshals it into the target structure "into".
+// Closes the input io.ReadCloser once done.
+func ReadSanitizeConfigJSON(headers http.Header, input io.ReadCloser, into interface{}) error {
+	// Performs the Content-Type validation and Content-Encoding check.
+	input, err := processContentEncoding(headers, input)
+	if err != nil {
+		return err
+	}
 
-	_ = input.Close()
+	content, err := ioutil.ReadAll(input)
+	if err != nil {
+		return err
+	}
+	defer func() { _ = input.Close() }()
+
+	// Expand environment variables.
+	content = expandEnv(content)
+
+	// Convert the back quotes into double-quotes, escapes literal
+	// backslashes, newlines or double-quotes with backslashes.
+	if content = sanitizeConfig(content); err != nil {
+		return err
+	}
 
+	// Decode the body bytes into target structure.
+	err = decode(content, into)
+	if err != nil {
+		err = base.WrapJSONUnknownFieldErr(err)
+		if errors.Cause(err) == base.ErrUnknownField {
+			err = base.HTTPErrorf(http.StatusBadRequest, "JSON Unknown Field: %s", err.Error())
+		} else {
+			err = base.HTTPErrorf(http.StatusBadRequest, "Bad JSON: %s", err.Error())
+		}
+	}
 	return err
 }
 
+// processContentEncoding performs the Content-Type validation and Content-Encoding check.
+func processContentEncoding(headers http.Header, input io.ReadCloser) (io.ReadCloser, error) {
+	contentType := headers.Get("Content-Type")
+	if contentType != "" && !strings.HasPrefix(contentType, "application/json") {
+		return input, base.HTTPErrorf(http.StatusUnsupportedMediaType, "Invalid content type %s", contentType)
+	}
+	switch headers.Get("Content-Encoding") {
+	case "gzip":
+		var err error
+		if input, err = gzip.NewReader(input); err != nil {
+			return input, err
+		}
+	case "":
+		break
+	default:
+		return input, base.HTTPErrorf(http.StatusUnsupportedMediaType, "Unsupported Content-Encoding; use gzip")
+	}
+	return input, nil
+}
+
+// sanitizeConfig converts the back quotes into double-quotes, escapes
+// literal backslashes, newlines or double-quotes with backslashes.
+func sanitizeConfig(content []byte) []byte {
+	content = base.ConvertBackQuotedStrings(content)
+	return content
+}
+
+// decode unmarshals the raw JSON bytes into the target structure. Underlying
+// Decoder to return an error when the destination is a struct and the input
+// contains object keys which do not match any non-ignored, exported fields
+// in the destination. UseNumber is enforced in Decoder to unmarshal a number
+// into an interface{} as a Number instead of as a float64.
+func decode(content []byte, into interface{}) error {
```

Comment:
This is kind of a vauge name, and the function signature causes it to be too limiting (it forces the caller to drain and rebuffer an io.Reader), whereas previously we could stream a HTTP Request body directly to the JSON decoder.

If the painful part is to set DisallowUnknownFields and UseNumber on a JSONDecoder, how about we create a new function in `base` that returns a decoder configured for these?

And then usage of it reduces down to
```golang
err := base.JSONDecoderWithOptions(r, true, true).Into(&into)
```

```golang
// JSONDecoderWithOptions returns a new JSON decoder implementing the JSONDecoderI interface with the specified options.
func JSONDecoderWithOptions(r io.Reader, disallowUnknownFields, useNumber bool) (d JSONDecoderI) {
	d := JSONDecoder(r)
	if disallowUnknownFields {
		d.DisallowUnknownFields()
	}
	if useNumber {
		d.UseNumber()
	}
	return d
}
```

Reply:
This makes more sense and it can be reused in both the places.